core/src/common.cpp

```diff
@@ -362,6 +364,439 @@ namespace irods::http
+		req.set(beast::http::field::host, irods::http::create_host_field(url, *port));
+		req.set(beast::http::field::user_agent, irods::http::version::server_name);
+		req.set(beast::http::field::accept, "application/json");
+		req.prepare_payload();
+
+		// Send request & receive response
+		auto res{tcp_stream->communicate(req)};
+
+		logging::debug("{}: Received the following response: [{}]", __func__, res.body());
+
+		return res.body();
+	}
+
+	/// Adds the specified symmetric algorithm \p _alg to the verifier \p _verifier, using the secrets provided in the
+	/// HTTP API configuration
+	///
+	/// See RFC 7518 for details on JSON Web Algorithms (JWA)
+	///
+	/// \param[in,out] _verifier The jwt::verifier to add additional verification algorithms to.
+	/// \param[in]     _alg      The sigining algorithm requested by the signed JWT.
+	auto add_symmetric_alg(
+		jwt::verifier<jwt::default_clock, jwt::traits::nlohmann_json>& _verifier,
+		std::string_view _alg) -> void
+	{
+		namespace logging = irods::http::log;
+
+		auto algorithm_family{_alg.substr(0, 2)};
+		auto algorithm_specifics{_alg.substr(2)};
+
+		// Extract the secret key
+		// OpenID secret should be our secret
+		std::string key;
+		if (auto realm_secret{irods::http::globals::oidc_configuration().find("realm_secret")};
+		    realm_secret != std::end(irods::http::globals::oidc_configuration()))
+		{
+			key = realm_secret->get<std::string>();
+		}
+		else if (auto secret{irods::http::globals::oidc_configuration().find("client_secret")};
+		         secret != std::end(irods::http::globals::oidc_configuration()))
+		{
+			key = secret->get<std::string>();
+		}
+
+		if (algorithm_family == "HS") {
+			if (algorithm_specifics == "256") {
+				logging::trace("{}: Adding [{}] to allowed verification algorithms.", __func__, _alg);
+				_verifier.allow_algorithm(jwt::algorithm::hs256(key));
+			}
+			else if (algorithm_specifics == "384") {
+				logging::trace("{}: Adding [{}] to allowed verification algorithms.", __func__, _alg);
+				_verifier.allow_algorithm(jwt::algorithm::hs384(key));
+			}
+			else if (algorithm_specifics == "512") {
+				logging::trace("{}: Adding [{}] to allowed verification algorithms.", __func__, _alg);
+				_verifier.allow_algorithm(jwt::algorithm::hs512(key));
+			}
+			return;
+		}
+
+		logging::warn("{}: Algorithm [{}] is not supported.", __func__, _alg);
+	}
+
+	/// Adds the specified asymmetric algorithm \p _alg to the verifier \p _verifier, using the additional information
+	/// provided by the JWK \p _jwk.
+	///
+	/// See RFC 7518 for details on JSON Web Algorithms (JWA)
+	///
+	/// \param[in,out] _verifier The jwt::verifier to add additional verification algorithms to.
+	/// \param[in]     _jwk      The JWK containing the required JWA information.
+	/// \param[in]     _alg      The signing algorithm requested by the signed JWT.
+	auto add_asymmetric_alg_from_jwk(
```

Comment:
Currently, if `add_asymmetric_alg_from_jwk(...)` is passed an invalid JWK such that `.get_jwk_claim(...)` fails, the exception will bubble up and prevent any further action being done on the JWT to be validated. Is this desired?

I think so, but I'll leave this here to think about it later.

Reply:
> Is this desired?

What are the pros and cons of letting it bubble up?
What does the admin see?
What does the client see?
What should the client do if they provide an invalid JWK?

Reply:
> What are the pros and cons of letting it bubble up?

The processing of the JWT stops, as it should imo, as the JWK is invalid.
This means there is likely an issue with the OpenID provider.

The cons are that there may be another JWK that is valid?

> What does the admin see?

The admin should see a complaint about an invalid JWK (logs).

> What does the client see?

Client sees an auth failure.

> What should the client do if they provide an invalid JWK?

The client does not provide the JWK, the OpenID Provider does, when querying the well-known endpoint.

Reply:
> This means there is likely an issue with the OpenID provider.
>
> The cons are that there may be another JWK that is valid?

So if there's an issue with the OIDC provider, then there's no reason to worry about another JWK being valid, correct?

Is this guaranteed to be an issue with the OIDC provider?

Reply:
Okay. Letting it ride until we learn more.

Please mark as resolved if you feel there's nothing to address and the implementation is good.

Reply:
Resolving.